### ISSUE TITLE
Root pattern for Gradle projects?

### ISSUE THREAD

**Issue by HiPhish** [neovim/nvim-lspconfig#455]
[Gradle](https://gradle.org/) is a build system used by JVM languages such as Java or Kotlin. A Gradle project can bei either a plain project or contain sub-project. The question is, what should the root pattern be?

A plain project will have one file `build.gradle` in its root directory. A project with sub-projects on the other hand will have multiple sub-directories, each with their own `build.gradle`, and one global `settings.gradle` file. In order to find the root directory we have to first search for a `settings.gradle` file, and only then fall back to searching for `build.gradle`. Otherwise the sub-project will be considered the root of the entire project.

My question is, what should the `root_dir` configuration be? At first I tried the following:

```lua
util.root_pattern('settings.gradle', 'build.gradle')
```

However, after looking at the implementation of `root_pattern` I can see that this is wrong. Neovim will for each directory (starting with the one containing the current file) look for all the root files first before moving up a directory. What I want is the opposite though: start with the first file, search upwards, and when that fails search for the second file.

Here is another attempt:

```lua
function(startpath)
    return util.root_pattern('settings.gradle')(startpath) or util.root_pattern('build.gradle')(startpath)
end
```

What do you think? Should we have a depth-first alternative for `util.root_path`? Or do we leave it to server configuration s to use the latter?

**Comment by HiPhish** [comment]
I slept a bit over it, and I came up with the following function:

```lua
local function root_patterns(...)
	local searchers = {}

	for _, patterns in ipairs({...}) do
		local searcher = root_pattern(unpack(patterns))
		searchers[#searchers] = searcher
	end

	return function(startpath)
		for _, searcher in ipairs(searchers) do
			local root = searcher(startpath)
			if root then return root end
		end
	end
end
```

We can then use it as follows:

```lua
root_dir = root_patterns({'settings.gradle', 'settings.gradle.kts'}, {'build.gradle', 'build.gradle.kts'})
```

As you can see, each of the arguments is a list of patterns. The reason for that is that I want to be able to specify several patters to search for in the same iteration. For example, I know that in a project with sub-projects there has to be one of two files: `settings.gradle` (written in [Groovy](https://groovy-lang.org/)) or `settings.gradle.kts` (written in [Kotlin](https://kotlinlang.org/)). There is no point in doing a depth-first search for one of them and then the other, when I could just execute a depth-first search for both at the same time.

**Comment by mizlan** [comment]
> There is no point in doing a depth-first search for one of them and then the other, when I could just execute a depth-first search for both at the same time.

Not sure how much of a concern this is, with regard to time. Otherwise this is similar to #237 .

**Comment by HiPhish** [comment]
@mizlan Yes, it sounds like the same issue to me. Anyway, I have been using my snippet above succesfully for a while now and it works at least for me (after fixing an emberrasing bug).

> Not sure how much of a concern this is, with regard to time

What exactly do you mean?

**Comment by mjlbach** [comment]
I feel like this is a good candidate for local project configuration with .nvimrc, so you can set whatever project root you want without fiddling with the root pattern.

**Comment by HiPhish** [comment]
@mjlbach You mean by putting a `.nvimrc` file in the project directory that sets the configuration of the project? I think that should be a last resort for when someone has an unusual project structure. A multi-module project is a completely standard thing to do in Gradle and in my opinion this should work out of the box.

**Comment by mjlbach** [comment]
In that case, does: 

```
function(startpath)
    return util.root_pattern('settings.gradle')(startpath) or util.root_pattern('build.gradle')(startpath)
end
```

Not provide the correct directory?  I'd like to avoid adding a bunch of handlers and prefer to show how to compose these. The actual search time is incredibly short.

**Comment by HiPhish** [comment]
Yes, that would do it. My idea was to have a higher-level `util` function which can be re-used, otherwise one has to type out that pattern every single time (we now have two servers where this is necessary). That's my proposal at least, I leave the decision to the repo maintainers.

**Comment by mjlbach** [comment]
Yeah, I totally sympathize with that idea as well. My completely very non-set opinion is that we should just add this workaround (for now), and then look through the root patterns in 6 months or so once we have a number of servers, and see what root patterns would actually benefit from a more general utility function.

If it's only 2/60 odd servers, not worth adding a separate function. If it ends up being 100/300, then probably worth some refactoring :)

**Comment by ahmedelgabri** [comment]
> If it's only 2/60 odd servers, not worth adding a separate function. If it ends up being 100/300, then probably worth some refactoring :)

IMO, it's not about the number of servers as much as that the expectation from `root_pattern` is that it checks files in order. _I might argue_ that the default behaviour is not right.

My naive assumption is that many people might be having issues with their config or think that it doesn't work because of that, without even noticing. nvim-lspconfig should have sane defaults & IMO one of them is that `root_pattern` should search in order of the list provided (it was definitely a surprising behaviour for me that it didn't). And document the behaviour so anyone can modify it if needed. 

This is by not criticism, it's just my 2 cents on the issue.

**Comment by HiPhish** [comment]
@ahmedelgabri `root_pattern` does check files in order, the issue is how it traverses directories. Starting with the directory of the file it checks for the patterns in the order they are given, then it moves up one directory, checks the patterns in the order they are given, moves up one directory, and so one until one pattern matches.

Thus if you have a list like `{foo, bar, baz}` and a directory structure like

```
├── foo
└── source
    └── bar
```
It will match `source/bar`. If there was also a file `source/foo`, then that would have been mached before `source/bar` because `foo` comes before `bar`.

**Comment by ahmedelgabri** [comment]
Thanks, @HiPhish, yeah I meant the way it resolves them. I made a mistake there.

**Comment by mjlbach** [comment]
Hi all, please give feedback on #671 if you have the time

**Comment by HiPhish** [comment]
> Hi all, please give feedback on #671 if you have the time

Looks good to me.

**Comment by mjlbach** [comment]
Closed via #370
